pkg/tfcli/refresh.go

```diff
@@ -19,136 +19,98 @@ package tfcli
 import (
 	"context"
 	"fmt"
-	"path/filepath"
-	"regexp"
-	"strconv"
 	"strings"
 
 	"github.com/pkg/errors"
 	"go.uber.org/multierr"
 
+	"github.com/crossplane-contrib/terrajet/pkg/json"
 	"github.com/crossplane-contrib/terrajet/pkg/tfcli/model"
 )
 
 const (
-	// error messages
-	errRefresh        = "failed to refresh the Terraform state"
-	errNoID           = errRefresh + ": cannot deduce resource ID"
-	errImport         = "failed to import resource"
-	fmtErrRefreshExit = errRefresh + ": Terraform pipeline exited with code: %d"
-	fmtErrImport      = errImport + ": %s"
+	errFmtObserve   = "observe failed with the following error: %s"
+	errFmtNoPlan    = "plan line not found in Terraform CLI output: %s"
+	errFmtParsePlan = "failed to parse Terraform plan output: %s"
 )
 
 // Refresh updates local state of the Cloud resource in a synchronous manner
-func (c *Client) Refresh(ctx context.Context, id string) (model.RefreshResult, error) {
-	if id == "" && len(c.tfState) == 0 {
-		return model.RefreshResult{}, errors.New(errNoID)
-	}
-
+func (c *Client) Refresh(ctx context.Context, _ string) (model.RefreshResult, error) {
 	if _, err := c.initConfiguration(model.OperationRefresh, false); err != nil {
 		return model.RefreshResult{}, err
 	}
-
-	var result model.RefreshResult
-	var err error
-	if len(c.tfState) == 0 {
-		result, err = c.importResource(ctx, id)
-	} else {
-		result, err = c.observe(ctx)
-	}
-	return result, multierr.Combine(err, c.removeStateStore())
-}
-
-func (c *Client) importResource(ctx context.Context, id string) (model.RefreshResult, error) {
-	result := model.RefreshResult{}
-	// as a precaution, remove any existing state file
-	if err := c.fs.RemoveAll(filepath.Join(c.wsPath, fileState)); err != nil {
-		return result, errors.Wrap(err, errImport)
-	}
-	// now try to run the synchronous import pipeline
-	if err := c.syncPipeline(ctx, true, pathTerraform, "import", "-input=false", c.resource.GetAddress(), id); err != nil {
+	// Prepare the workspace for a refresh operation because the state that comes
+	// with import may not include all information we have. So, we first write
+	// everything into state file and then run refresh to update the ones that
+	// can be updated with the current state of the world. Import cannot be ru
+	// with existing state file.
```

Comment:
Did we lose the import functionality?

Reply:
Hmm, actually we didn't, we import just via refreshing state with enough fields.

Reply:
Yes we still have the import functionality. Since we removed the annotation, we don't have a signal as clear as `len(c.tfState) == 0` to decide whether we need to call import or not. Checking whether observation or parameters are empty do not always mean that the resource is not created or not in creating state because it could take a few reconciles until the creation is completed and observation reflects the state.

Reply:
Extended the comment to describe why we use refresh all the time https://github.com/crossplane-contrib/terrajet/pull/72/commits/cdf118898f9afa211b773e86a6c5765ceaffc3a6